### DISCUSSION TITLE
源码中的混合模式求助

### DISCUSSION THREAD

**Discussion by Rey5331** [Tencent/libpag#1574] [Q&A] [answered]
为了拓展其它混合模式 我观察 tgfx/src/gpu/opengl/GLBlend.cpp 文件中的一些混合模式 但发现和我目前的写法不太一样 我想知道它们为什么是这样的 是先预乘了alpha 然后又把之后要做的mix都写进来了吗？部分代码如下(我的代码最后都会做一个mix(src.rgb, output.rgb, output.a))：
mutiply 	: 
	libpag 	: output.rgb = (1.0 - src.a) * dst.rgb + (1.0 - dst.a) * src.rgb + src.rgb * dst.rgb
	my		: output.rgb = src.rgb * dst.rgb

darken 		: 
	libpag	: output.rgb = min((1.0 - src.a) * dst.rgb + src.rgb, (1.0 - dst.a) * src.rgb + dst.rgb)
	my		: min(dst.rgb, src.rgb)

difference 	: 
	libpag	: output.rgb = src.rgb + dst.rgb - 2.0 * min(src.rgb * dst.a, dst.rgb * src.a)
	my		: output.rgb = abs(src.rbg - dst.rgb) = src.rgb + dst.rgb - 2.0 * min(src.rgb, dst.rgb)

exclusion	: 
	libpag	: output.rgb = dst.rgb + src.rgb - 2.0 * dst.rgb * src.rgb
	my		: output.rgb = dst.rgb + src.rgb - 2.0 * dst.rgb * src.rgb

**Comment by domchen** [comment, marked as answer]
我们所有shader里的颜色输入都预乘过alpha的，不然存在纹理绘制的情况性能比较差。渲染引擎一般都是强制预乘Alpha，很少存在不预乘的情况。

**Reply by Rey5331** [reply]
原来如此 多谢

**Reply by Rey5331** [reply]
我想再请教一下 如何推导出和你们代码相同格式的公式
比如Divide混合模式 我认为没预乘alpha时 原理是
output = vec4(mix(dst.rgb, (dst.rgb / src.rgb).rgb, src.a), dst.a)
我用预乘的rgb推导出来的公式是
output.rgb = (1.0 - src.a) * dst.rgb + (1.0 - dst.a) * src.rgb + src.a * src.a * dst.rgb / src.rgb
 但并没有达到效果
我想学习预乘alpha的计算过程 如果还不行的话 我可能要用最蠢的办法 把预乘的rgb除回去 再mix了……

**Reply by Rey5331** [reply]
我解决了问题 最终公式是
output.rgb = (1.0 - src.a) * dst.rgb + (1.0 - dst.a) * src.rgb + src.a * clamp(src.a * dst.rgb / src.rgb, 0.0, 1.0)
但我知道这样一点都没用上预乘的思想😢
